### PR TITLE
Show location for each peer

### LINKED ISSUE #1512: indication of country
is there any plan to add the name of the country or its flag to the peer list?

### PR DESCRIPTION
**Pull request by gmou3**
I wish to make `Transmission` show the location of each peer in the `DetailsDialog`.

I found a free database from [db-ip](https://db-ip.com/db/download/ip-to-city-lite), and created a repo that automatically updates the required `mmdb` database (which changes monthly) and transforms an IP to a location string `"City, Country"`. This repo contains a `C` file that needs to be build by linking with `libmaxminddb`. I made `transmission-gtk` automatically check the folder `$HOME/.ip-to-location` for the executable. I have only tested it on `Linux`.

So, in order to make things work, install `transmission-gtk` from source as usual, install `libmaxminddb` using your system's package manager, and run
```
git clone https://github.com/gmou3/ip-to-location $HOME/.ip-to-location
cd $HOME/.ip-to-location
gcc ip_to_location.c -o ip_to_location -libmaxminddb
```
(Note that even if you skip these extra steps everything works normally with an empty `Location` column.)

Then run `transmission-gtk` and go to the `Properties` dialog of a torrent, under `Peers` to see the changes. The first time will take some time as it downloads the `mmdb` database file (`~50MB`).

I suppose there is a more proper integration method, using `cmake` and building `libmaxminddb` as a `third-party` submodule. I tried a bit but couldn't achieve that. There also exist issues of compatibility outside `Linux`.
Any help will be much appreciated!

Resolves #1512.

### CODE REVIEW

**Review by ckerr — CHANGES_REQUESTED** [review]
- calling `system()` + `access()` + `fopen()` + `fgets()` + `fclose()` per peer seems prohibitively expensive to me.

- there are several potential buffer overruns here and portability problems e.g. requiring `getpwuid()` and `getuid()` to be present. Even if we did want to call an executable for each peer, using more portable API like `Glib::get_home_dir()` would be better

- This isn't a general-purpose solution because it requires each end user to download and compile a tool from GitHub. And I'm not positive, but also maybe register for a key from MaxMind to get access to their database? Either way, this is something the average user is never going to do.

tldr, this is an extremely heavyweight solution to a feature that doesn't call for it. I'd _much_ rather have something lightweight & embedded so that  end users don't have to do anything and so that it doesn't make Transmission slower.

**Comment by gmou3** [comment]
> * there are several potential buffer overruns here and portability problems e.g. requiring `getpwuid()` and `getuid()` to be present. Even if we did want to call an executable for each peer, using more portable API like `Glib::get_home_dir()` would be better

Thanks for the recommendation. I was able to avoid `include <pwd.h>`, but maybe the path construction after using `Glib::get_home_dir()` is still problematic outside Linux.

> * calling `system()` + `access()` + `fopen()` + `fgets()` + `fclose()` per peer seems prohibitively expensive to me.
> * This isn't a general-purpose solution because it requires each end user to download and compile a tool from GitHub. And I'm not positive, but also maybe register for a key from MaxMind to get access to their database? Either way, this is something the average user is never going to do.

No database key is necessary. The idea is to delegate to an external package the whole work of the db management (initial download and updates), as well as the individual calls for the IP to location functionality. We could internalize some of this functionality within Transmission, and we can avoid the checks for each peer (maybe check things only at startup). Internalizing the IP to location functionality would require having the libmaxmind library (or managing its absence). 

Any helping hands welcome.

**Comment by gmou3** [comment]
@ckerr after quite some work things are looking good. The functionality is embedded now and most of your initial points have been addressed. Please have a look at the updated PR post for details about the changes.

Some remaining things to be improved are the location of the `mmdb` file and, possibly, the making of `maintain_mmdb_file` asynchronous.

**Comment by oooo-ps** [comment]
To build this branch on the older Glib versions:

I've replaced:
``` gtk/IPToLocation.cc
if (!Glib::file_test(mmdb_file, Glib::FileTest::EXISTS))
```
to
``` gtk/IPToLocation.cc
if (!Glib::file_test(mmdb_file, Glib::FileTest::FILE_TEST_EXISTS))
```

Some [tips](https://devzone.org.ua/post/heo-lokatsiia-piriv-dlia-transmission-gtk) about how to build this branch on Fedora 42 (Uk).

**Comment by gmou3** [comment]
@oooo-ps thanks for your feedback!
Does it work smoothly for you, otherwise?

**Comment by oooo-ps** [comment]
@gmou3 I'm using this branch about few months without any issues,
about this my `Glib::FileTest::EXISTS` issue - it is not actual after upgrade to GNOME 48

Thanks for your work!

**Comment by killemov** [comment]
Maybe take a look at some other local options? https://www.google.com/search?q=free+embed+ip+address+country+database

**Comment by gmou3** [comment]
This PR is now ready from my perspective. I've put some effort into it so that everything is clear and well-structured. The newly introduced global mmdb handler and lock logic work quite nice. I think it will be fun to review.
